consul/watcher.go

```diff
@@ -209,6 +216,75 @@ func (w *Watcher) startUpstream(up api.Upstream) {
 	}()
 }
 
+func (w *Watcher) startUpstreamPreparedQuery(up api.Upstream, name string) {
+	w.log.Infof("consul: watching upstream for prepared_query %s", up.DestinationName)
+
+	u := &upstream{
+		LocalBindAddress: up.LocalBindAddress,
+		LocalBindPort:    up.LocalBindPort,
+		Name:             name,
+		Datacenter:       up.Datacenter,
+	}
+
+	if p, ok := up.Config["protocol"].(string); ok {
+		u.Protocol = p
+	}
+
+	interval := preparedQueryPollInterval
+	if p, ok := up.Config["poll_interval"].(string); ok {
+		dur, err := time.ParseDuration(p)
+		if err != nil {
+			w.log.Errorf(
+				"consul: upstream %s %s: invalid poll interval %s: %s",
+				up.DestinationType,
+				up.DestinationName,
+				p,
+				err,
+			)
+			return
+		}
+		interval = dur
+	}
+
+	w.lock.Lock()
+	w.upstreams[name] = u
+	w.lock.Unlock()
+
+	go func() {
+		var last []*api.ServiceEntry
+		for {
+			if u.done {
+				return
+			}
+			nodes, _, err := w.consul.PreparedQuery().Execute(up.DestinationName, &api.QueryOptions{
+				Connect:    true,
+				Datacenter: up.Datacenter,
+				WaitTime:   10 * time.Minute,
+			})
+			if err != nil {
+				w.log.Errorf("consul: error fetching service definition for service %s: %s", up.DestinationName, err)
+				time.Sleep(errorWaitTime)
```

Comment:
Can you do a truncated binary exponential backoff instead of a fixed one ?

Reply:
I can, however, that's how errors are handled in all the other watches. I can make this a separate PR.